### ISSUE TITLE
oc adm migrate {storage, etcd-ttl, legacy-hpa} does not display all available options

### ISSUE THREAD

**Issue by nak3** [openshift/origin#19083]
- current (only `--filename` is completed):
~~~
$ oc adm migrate storage -<TAB> <TAB>
-f           --filename=
~~~

- expected (all available options should be completed.):
~~~
$ oc adm migrate storage -<TAB><TAB>
--all-namespaces               --client-certificate=          --filename=                    --logspec=                     --server=                      --to-key=
--allow-missing-template-keys  --client-key=                  --from-key=                    --match-server-version         --show-all                     --user=
  (snip)
~~~

As told by @env in https://github.com/openshift/origin/pull/19046#issuecomment-375703584, I am opening this issue. I previously created https://github.com/openshift/origin/pull/19046, but it was told as wrong approach.

This is not a big issue, but I really would like to learn:

- Why `MarkFlagRequired` is enabled for `filename` option in `oc adm migrate storage` even though not mandatory.
- How https://github.com/spf13/cobra/issues/407 was related to this issue.

as mentioned in https://github.com/openshift/origin/pull/19046.

**Comment by jwforres** [comment]
@openshift/sig-master

**Comment by juanvallejo** [comment]
@enj 
> Why MarkFlagRequired is enabled for filename option in oc adm migrate storage even though not mandatory.

Do you know if there is a specific reason to keep `--filename` a required flag in this command?

@nak3 
> How spf13/cobra#407 was related to this issue.

Just saw [your comment](https://github.com/openshift/origin/pull/19046#issuecomment-375721648), both bugs seem to describe the same issue to me. The bug you are describing is not being able to see more than the flag marked as required when you attempt to do tab-completion?

Additionally, although removing MarkFlagRequired does appear to fix this bug, it does not address the underlying cause, which is that tab completions do not work as expected when there are flags marked as required.

**Comment by liggitt** [comment]
once the require flag is provided, the other flags tab-complete

fwiw, there are two flags ("filename" arg of this command, and the "dest-ca-cert" arg of the "create route" command) I stopped marking as required in the 1.10 rebase in progress (spf13 got even stricter about flags marked as required in the new version we pick up in 1.10, and actually refuses to run the command)

**Comment by nak3** [comment]
> The bug you are describing is not being able to see more than the flag marked as required when you attempt to do tab-completion?

No, it is different.  Actually it is possible to see available flags, if I provided a dummy value for `--filename` as @liggitt also mentioned on bash 4. But `--filename` is not required for `oc adm migrate`. (So, I wanted to remove `MarkFlagRequired` for `--filenmae` by https://github.com/openshift/origin/pull/19046 )

```
$ oc adm migrate storage --filename dummy -[tab]
--all-namespaces            --client-certificate=       --confirm                   --from-key=                 --logspec=                  --no-headers                --server=
--as=                       --client-key=               --context=                  --include=                  --match-server-version      -o                          --token=
--as-group=                 --cluster=                  -f                          --insecure-skip-tls-verify  -n                          --output=                   --to-key=
--certificate-authority=    --config=                   --filename=                 --loglevel=                 --namespace=                --request-timeout=          --user=
```

**Comment by juanvallejo** [comment]
@nak3 makes sense. I'm fine with removing MarkFileRequired for --filename.
@enj any objections to this?

**Comment by liggitt** [comment]
already done in https://github.com/openshift/origin/pull/19200/commits/4385e4047982999df0111df16f2783a87acc62f3
